File: src/main/java/ru/arc/invest/items/NamedItem.java
package ru.arc.invest.items;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

import lombok.AllArgsConstructor;
import lombok.ToString;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.inventory.ItemStack;
import ru.arc.ARC;

@AllArgsConstructor
@ToString
public class NamedItem extends GenericItem {

    private static final Map<String, NamedItem> namedItemMap = new HashMap<>();

    private ItemStack item;
    private String name;

    public static NamedItem find(String name) {
        return namedItemMap.get(name.toLowerCase());
    }

    public static NamedItem deserialize(Map<String, Object> map, String id) {
        ItemStack stack = ItemStack.deserialize(map);
        return new NamedItem(stack, id);
    }

    public static void load() {
        namedItemMap.clear();
        Path path = Paths.get(ARC.getInstance().getDataFolder().toString(), "investing", "items");
        if (!Files.exists(path)) {
            try {
                Files.createDirectories(path);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        try (var stream = Files.walk(path, 3)) {
            stream.filter(p -> p.toString().endsWith(".yml"))
                    .map(Path::toFile)
                    .map(YamlConfiguration::loadConfiguration)
                    .forEach(NamedItem::readFile);
        } catch (Exception e) {
            ARC.getInstance().getLogger().warning("Error loading named items: " + e.getMessage());
        }

        System.out.println("Loaded named items: " + namedItemMap);
    }

    private static void readFile(YamlConfiguration configuration) {
        for (String key : configuration.getKeys(false)) {
            Object val = configuration.get(key);
            if (!(val instanceof Map<?, ?> rawMap)) {
                continue;
            }
            @SuppressWarnings("unchecked")
            Map<String, Object> typedMap = (Map<String, Object>) rawMap;
            namedItemMap.put(key.toLowerCase(), NamedItem.deserialize(typedMap, key));
        }
    }

    @Override
    public ItemStack stack() {
        return item.clone();
    }

    @Override
    public boolean fits(ItemStack stack) {
        return item.equals(stack.asOne());
    }
}

File: src/main/kotlin/ru/arc/invest/items/NamedItem.kt
package ru.arc.invest.items

import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import ru.arc.ARC
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

class NamedItem(private val item: ItemStack, private val name: String) : GenericItem() {

    override fun stack(): ItemStack = item.clone()

    override fun fits(stack: ItemStack): Boolean = item == stack.asOne()

    override fun toString(): String = "NamedItem{name=$name}"

    companion object {
        private val namedItemMap = HashMap<String, NamedItem>()

        @JvmStatic
        fun find(name: String): NamedItem? = namedItemMap[name.lowercase()]

        @JvmStatic
        fun deserialize(map: Map<String, Any>, id: String): NamedItem {
            val stack = ItemStack.deserialize(map)
            return NamedItem(stack, id)
        }

        @JvmStatic
        fun load() {
            namedItemMap.clear()
            val path = Paths.get(ARC.instance.dataFolder.toString(), "investing", "items")
            if (!Files.exists(path)) {
                try {
                    Files.createDirectories(path)
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            }

            try {
                Files.walk(path, 3).use { stream ->
                    stream.filter { it.toString().endsWith(".yml") }
                        .map { it.toFile() }
                        .map { YamlConfiguration.loadConfiguration(it) }
                        .forEach { readFile(it) }
                }
            } catch (e: Exception) {
                ARC.instance.logger.warning("Error loading named items: ${e.message}")
            }

            println("Loaded named items: $namedItemMap")
        }

        @Suppress("UNCHECKED_CAST")
        private fun readFile(configuration: YamlConfiguration) {
            for (key in configuration.getKeys(false)) {
                val value = configuration.get(key)
                if (value !is Map<*, *>) continue
                val typedMap = value as Map<String, Any>
                namedItemMap[key.lowercase()] = deserialize(typedMap, key)
            }
        }
    }
}
